### PR TITLE
8284742: x86: Handle integral division overflow during parsing

### PR DESCRIPTION
**Pull request by merykitty**
Hi,

This patch moves the handling of integral division overflow on x86 from code emission time to parsing time. This allows the compiler to perform more efficient transformations and also aids in achieving better code layout.

I also removed the handling for division by 10 in the ad file since it has been handled in `DivLNode::Ideal` already.

Thank you very much.

    Before:
    Benchmark                               (BUFFER_SIZE)  (divisorType)  Mode  Cnt      Score     Error  Units
    IntegerDivMod.testDivide                         1024          mixed  avgt    5   2394.609 ±  66.460  ns/op
    IntegerDivMod.testDivide                         1024       positive  avgt    5   2411.390 ± 136.849  ns/op
    IntegerDivMod.testDivide                         1024       negative  avgt    5   2396.826 ±  57.079  ns/op
    IntegerDivMod.testDivideHoistedDivisor           1024          mixed  avgt    5   2121.708 ±  17.194  ns/op
    IntegerDivMod.testDivideHoistedDivisor           1024       positive  avgt    5   2118.761 ±  10.002  ns/op
    IntegerDivMod.testDivideHoistedDivisor           1024       negative  avgt    5   2118.739 ±  22.626  ns/op
    IntegerDivMod.testDivideKnownPositive            1024          mixed  avgt    5   2467.937 ±  24.213  ns/op
    IntegerDivMod.testDivideKnownPositive            1024       positive  avgt    5   2463.659 ±   6.922  ns/op
    IntegerDivMod.testDivideKnownPositive            1024       negative  avgt    5   2480.384 ± 100.979  ns/op

    Benchmark                               (BUFFER_SIZE)  (divisorType)  Mode  Cnt      Score     Error  Units
    LongDivMod.testDivide                            1024          mixed  avgt    5   8312.558 ±  18.408  ns/op
    LongDivMod.testDivide                            1024       positive  avgt    5   8339.077 ± 127.893  ns/op
    LongDivMod.testDivide                            1024       negative  avgt    5   8335.792 ± 160.274  ns/op
    LongDivMod.testDivideHoistedDivisor              1024          mixed  avgt    5   7438.914 ±  17.948  ns/op
    LongDivMod.testDivideHoistedDivisor              1024       positive  avgt    5   7550.720 ± 572.387  ns/op
    LongDivMod.testDivideHoistedDivisor              1024       negative  avgt    5   7454.072 ±  70.805  ns/op
    LongDivMod.testDivideKnownPositive               1024          mixed  avgt    5  12120.874 ±  82.832  ns/op
    LongDivMod.testDivideKnownPositive               1024       positive  avgt    5   8898.518 ±  29.827  ns/op
    LongDivMod.testDivideKnownPositive               1024       negative  avgt    5    562.742 ±   2.795  ns/op

    After:
    Benchmark                               (BUFFER_SIZE)  (divisorType)  Mode  Cnt      Score     Error  Units
    IntegerDivMod.testDivide                         1024          mixed  avgt    5   2174.521 ±  13.054  ns/op
    IntegerDivMod.testDivide                         1024       positive  avgt    5   2172.389 ±   7.721  ns/op
    IntegerDivMod.testDivide                         1024       negative  avgt    5   2171.290 ±  12.902  ns/op
    IntegerDivMod.testDivideHoistedDivisor           1024          mixed  avgt    5   2049.926 ±  29.098  ns/op
    IntegerDivMod.testDivideHoistedDivisor           1024       positive  avgt    5   2043.896 ±  11.702  ns/op
    IntegerDivMod.testDivideHoistedDivisor           1024       negative  avgt    5   2045.430 ±  17.232  ns/op
    IntegerDivMod.testDivideKnownPositive            1024          mixed  avgt    5   2281.506 ±  81.440  ns/op
    IntegerDivMod.testDivideKnownPositive            1024       positive  avgt    5   2279.727 ±  21.590  ns/op
    IntegerDivMod.testDivideKnownPositive            1024       negative  avgt    5   2275.898 ±   3.692  ns/op

    Benchmark                               (BUFFER_SIZE)  (divisorType)  Mode  Cnt      Score     Error  Units
    LongDivMod.testDivide                            1024          mixed  avgt    5   8321.347 ±  93.932  ns/op
    LongDivMod.testDivide                            1024       positive  avgt    5   8352.279 ± 213.565  ns/op
    LongDivMod.testDivide                            1024       negative  avgt    5   8347.779 ± 203.612  ns/op
    LongDivMod.testDivideHoistedDivisor              1024          mixed  avgt    5   7313.156 ± 113.426  ns/op
    LongDivMod.testDivideHoistedDivisor              1024       positive  avgt    5   7299.939 ±  38.591  ns/op
    LongDivMod.testDivideHoistedDivisor              1024       negative  avgt    5   7313.142 ± 100.068  ns/op
    LongDivMod.testDivideKnownPositive               1024          mixed  avgt    5   9322.654 ± 276.328  ns/op
    LongDivMod.testDivideKnownPositive               1024       positive  avgt    5   8639.404 ± 479.006  ns/op
    LongDivMod.testDivideKnownPositive               1024       negative  avgt    5    564.148 ±   6.009  ns/op

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8284742](https://bugs.openjdk.java.net/browse/JDK-8284742): x86: Handle integral division overflow during parsing

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)
 * [Martin Doerr](https://openjdk.java.net/census#mdoerr) (@TheRealMDoerr - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8206/head:pull/8206` \
`$ git checkout pull/8206`

Update a local copy of the PR: \
`$ git checkout pull/8206` \
`$ git pull https://git.openjdk.java.net/jdk pull/8206/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8206`

View PR using the GUI difftool: \
`$ git pr show -t 8206`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8206.diff">https://git.openjdk.java.net/jdk/pull/8206.diff</a>

</details>

### CODE REVIEW

**Comment by jatin-bhateja** [comment]
Hi @merykitty , Nice work!
Target specific IR generation looks interesting approach, but UDivL/UDivI are currently being generated by intrinsification route.  Thus a post parsing target lowering stage will ideally be suited. 

We can also take an alternative approach  to generate separate matcher rules for both the control paths by way of setting an attribute in IR node during Identity transformation.  
https://github.com/openjdk/jdk/pull/7572#discussion_r813918734

**Comment by merykitty** [comment]
@jatin-bhateja  Thanks a lot for your suggestions. The transformation manipulates the control flow so it should be handled during parsing since the control edge may have been lost right after that. The same goes for UDivL and UDivI intrinsic, too. I believe having target specific parsing is beneficial since we can decompose complex operations into more elemental ones, utilizing the power of the compiler more efficiently.

Delaying the handling till code emission time may miss the opportunities to hoist out the check and in the worst case would result in suboptimal code layout since the compiler can move the uncommon path out of the common path while the assembler can't.

**Review by vnkozlov — COMMENTED** [review]
I don't like that IR generation is placed in x86 platform specific code. Other platforms could also benefit from such IR (I checked, I'm only not sure about aarch64).

I suggest to do it in parse2.cpp by call a new method `parse_div_mod()` also defined in  this file. the method will have platform specific `Matcher::check_div_overflow()` (or something) to generate this special IR.

**Comment by jatin-bhateja** [comment]
> @jatin-bhateja Thanks a lot for your suggestions. The transformation manipulates the control flow so it should be handled during parsing since the control edge may have been lost right after that. The same goes for UDivL and UDivI intrinsic, too. I believe having target specific parsing is beneficial since we can decompose complex operations into more elemental ones, utilizing the power of the compiler more efficiently.
> 
> Delaying the handling till code emission time may miss the opportunities to hoist out the check and in the worst case would result in suboptimal code layout since the compiler can move the uncommon path out of the common path while the assembler can't.

Thanks I get your point, creating a control flow later may not be possible unless an explicit control projection is added to  DivI/L node during initial parsing which gets tied to its successors, and later on the IR node itself can be replaced by a control structure which converge at original control projection.  We can save redundant duplications of byte code processing and a later stage can do target lowering.

**Comment by merykitty** [comment]
@vnkozlov  I agree that this situation is common enough so I moved the logic to construct the branches to share code. I kept `Matcher::parse_one_bytecode` however since it would be helpful for other operations such as masking of shift nodes or handling of out-of-bounds floating-point-to-integer conversions.

@jatin-bhateja  The div nodes themselves are constructed with control input, it is the `Ideal` method that may remove this input. Post parse transformation does not only need to replace the div node with the control structure, but also has to restructure the control flow itself, so I think it would be much more complex to do so.

Thank you very much.

**Comment by vnkozlov** [comment]
Thanks, @merykitty

This is better. I am still not comfortable about passing `Parse` phase pointer to `Matcher` phase (they should not have such relation). May be create `Parse` specific file for platform.

I would like other platforms experts to review these changes and propose corresponding changes for their platforms.

@TheRealMDoerr, @RealLucy and @reinrich, please, review these changes. It looks like these platforms may benefit from this optimization too.

Expert from ARM, @nick-arm, and RISC-V, @RealFYang, please, look too.

Thanks!

**Comment by merykitty** [comment]
@vnkozlov Yes you are right, having the function in `Parse` instead of `Matcher` makes a lot more sense! Thanks for pointing it out.

**Comment by nick-arm** [comment]
On AArch64 we don't need the special case handling because the `sdiv` instruction already gives 0x80000000 / -1 = 0x8000000 and doesn't trap.

**Comment by jatin-bhateja** [comment]
> @jatin-bhateja The div nodes themselves are constructed with control input, it is the `Ideal` method that may remove this input. Post parse transformation does not only need to replace the div node with the control structure, but also has to restructure the control flow itself, so I think it would be much more complex to do so.
> 
> Thank you very much.

Hi @merykitty , Yes, input controlling edge cannot be used independently to create a control flow, we also need to have knowledge of next control region to create a conditional graph, thus adding an explicit output control projection from DivI/L IR nodes (which only have data flow edges) during parsing will enable creation of control flow later. Generating a control flow post parsing is not new we do it during macro expansion currently.

**Comment by merykitty** [comment]
@nick-arm Thanks a lot for your response. IIUC, AArch64 may benefit from this patch by parsing `irem` and `lrem` as `Sub src1 (Mul src2 (Div src1 src2))`, which would help eliminate a redundant `sdiv` instruction when both the remainder and division operations appear together.

File: src/hotspot/share/opto/parse3.cpp
Comment (TheRealMDoerr) [inline src/hotspot/share/opto/parse3.cpp:439]: One "the" too much.

File: test/micro/org/openjdk/bench/java/lang/IntegerDivMod.java
Comment (TheRealMDoerr) [inline test/micro/org/openjdk/bench/java/lang/IntegerDivMod.java:119]: Better add newline at the end.

File: test/micro/org/openjdk/bench/java/lang/LongDivMod.java
Comment (TheRealMDoerr) [inline test/micro/org/openjdk/bench/java/lang/LongDivMod.java:119]: Better add newline at the end.

**Review by TheRealMDoerr — COMMENTED** [review]
Thanks for pointing me to it! I think it makes sense for PPC64, too, but we can use separate PRs for other platforms (unless you prefer to wait).
I found some minor nits while looking at it.

**Review by RealLucy — COMMENTED** [review]
This is an interesting enhancement, and s390 would probably benefit as well. In line with what Martin said re. PPC64, I would suggest to handle the s390 implementation in a separate PR. I will not be able to invest enough time short-term.

**Comment by vnkozlov** [comment]
Thank you all for responses.

Changes seem fine to me now. I will run testing.

**Comment by jatin-bhateja** [comment]
> @jatin-bhateja Thank you very much for the explanations, I think I will go with this approach because it is much simpler. For your concern regarding the expansions of intrinsic nodes, I think we could have another arch-dependent parsing of intrinsic if it is needed, which has the benefits of being more general since intrinsic is often expanded into multiple nodes. By the way, `UDiv` and `UMod` do not need target lowering since unsigned division cannot overflow.
> 
@merykitty , UDiv/UMod also has a control flow, and as a generic rule bringing any control flow logic out from emit block into IR will get benefited from branch predictions based on profiles. In the best case some control paths may never be JIT compiled and would lead to uncommon traps. your approach to bring conditional logic to IR looks good to me.

**Comment by merykitty** [comment]
I have reworked the patch a little bit. A node must have consistent behaviour, as a result, we cannot just use `DivINode` and the likes as it would have inconsistent behaviour with the special case of integer division. As a result, we need other types of nodes which forbit this combination as inputs.

Thank you very much.

**Comment by theRealAph** [comment]
> @nick-arm Thanks a lot for your response. IIUC, AArch64 may benefit from this patch by parsing `irem` and `lrem` as `Sub src1 (Mul src2 (Div src1 src2))`, which would help eliminate a redundant `sdiv` instruction when both the remainder and division operations appear together.

It won't; we already do that.

**Comment by vnkozlov** [comment]
I agree that it should be specialized nodes. But the naming... It is easy to confuse with **no** (not).
I am not big expert in naming but maybe move verbose: `NoOvfDiv`.

I also see that you removed `divL_eReg_imm32` and `modL_eReg_imm32` from `x86_32.ad`. Please explain.

Previous changes passed my testing.

**Comment by merykitty** [comment]
@vnkozlov  I have renamed the nodes, thanks for your suggestions. The deletion was my mistake, I have redone that in the latest commit. Thank you very much.

**Comment by vnkozlov** [comment]
Good. I will test latest changes.

**Review by vnkozlov — APPROVED** [review]
Testing results are good.
You need second review.

**Comment by theRealAph** [comment]
This doesn't regress on AArch64, so I'm happy. For the record, my benchmark is this, which does BCD conversion by repeated division.

```
    public int base;
    public long nn = 0x124_6456_efab_8679l;

    @Benchmark
    public long BCDConversion() {
        nn++;
        long tmp = nn;
        long result = 0;
        for (int i = 0; i < 16; i++) {
            result = (result << 4) + tmp % base;
            tmp /= base;
        }
        return result;
    }

    @Benchmark
    public long unsignedBCDConversion() {
        nn++;
        long tmp = nn;
        long result = 0;
        for (int i = 0; i < 16; i++) {
            result = (result << 4) + Long.remainderUnsigned(tmp, base);
            tmp = Long.divideUnsigned(tmp, base);
        }
        return result;
    }

```

**Review by TheRealMDoerr — COMMENTED** [review]
This seems to be the first usage of lambda expressions in hotpot :-)
I'm ok with that and it is allowed for JDK19 AFAIK.

Using separate nodes makes sense. But, isn't it a problem that we still use "new DivINode" and "new DivLNode" (e.g. for loop transformations), but don't have match rules for them any more? I believe we need to use the NoOvf versions instead.

**Review by TheRealMDoerr — APPROVED** [review]
Ok, this makes sense. We only need to be careful when adding DivI/L nodes which don't get replaced.
Shared code looks good to me. I haven't checked the x86 code carefully.

**Comment by merykitty** [comment]
Thanks a lot for your reviews and discussions, do I need another review here? I would tentatively issue the integrate command in case it is not needed.
/integrate

**Pull request closed without merging**
